crates/doctor/src/rustlang/target.rs

```diff
@@ -0,0 +1,193 @@
+    WasmTargetNotInstalled,
+}
+
+impl Diagnosis for TargetDiagnosis {
+    fn description(&self) -> String {
+        match self {
+            Self::RustNotInstalled => "The Rust compiler isn't installed".into(),
+            Self::WasmTargetNotInstalled => {
+                "The required Rust target 'wasm32-wasi' isn't installed".into()
+            }
+        }
+    }
+
+    fn is_critical(&self) -> bool {
+        true
+    }
+
+    fn treatment(&self) -> Option<&dyn Treatment> {
+        Some(self)
+    }
+}
+
+#[async_trait]
+impl RustTreatment for TargetDiagnosis {
+    fn summary(&self) -> String {
+        match self {
+            Self::RustNotInstalled => "Install the Rust compiler and the wasm32-wasi target",
+            Self::WasmTargetNotInstalled => "Install the Rust wasm32-wasi target",
+        }
+        .into()
+    }
+
+    async fn treat_rust(&self) -> Result<()> {
+        match self {
+            Self::RustNotInstalled => {
+                install_rust().await?;
+                install_wasi_target()?;
+            }
+            Self::WasmTargetNotInstalled => {
+                install_wasi_target()?;
+            }
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl Treatment for TargetDiagnosis {
+    fn summary(&self) -> String {
+        RustTreatment::summary(self)
+    }
+
+    async fn dry_run(&self, _patient: &PatientApp) -> Result<String> {
+        let message = match self {
+            Self::RustNotInstalled => "Download and run the Rust installer from https://rustup.rs, then run `rustup target add wasm32-wasi`",
+            Self::WasmTargetNotInstalled => "Run the following command:\n    `rustup target add wasm32-wasi`",
+        };
+        Ok(message.into())
+    }
+
+    async fn treat(&self, _patient: &mut PatientApp) -> Result<()> {
+        self.treat_rust().await
+    }
+}
+
+async fn install_rust() -> Result<()> {
+    let status = run_rust_installer().await?;
+    anyhow::ensure!(status.success(), "Rust installation failed: {status:?}");
+    Ok(())
+}
+
+#[cfg(not(windows))]
+async fn run_rust_installer() -> Result<std::process::ExitStatus> {
+    use std::io::Write;
+
+    let resp = reqwest::get("https://sh.rustup.rs").await?;
+    let script = resp.bytes().await?;
+
+    let mut cmd = std::process::Command::new("sh");
+    cmd.stdin(std::process::Stdio::piped());
+    let mut shell = cmd.spawn()?;
+    let mut stdin = shell.stdin.take().unwrap();
+    std::thread::spawn(move || {
+        stdin.write_all(&script).unwrap();
+    });
+
+    let output = shell.wait_with_output()?;
+    Ok(output.status)
+}
+
+#[cfg(windows)]
+async fn run_rust_installer() -> Result<std::process::ExitStatus> {
+    // We currently distribute Windows builds only for x64, so hopefully
+    // this won't be an issue.
+    if std::env::consts::ARCH != "x86_64" {
+        anyhow::bail!("Spin Doctor can only install Rust for Windows on x64 processors");
+    }
```

Comment:
That makes `Diagnosis::treatment` more complicated, because it gets an OS/arch dependency in one case but not others.  It would be worth it if we had Windows builds other than x64, but we don't, so this is a "should never happen" scenario; and as such I think it's okay to put it in the place where the OS and treatment are already selected for.  Not that the extra complexity is enormous, so happy to do it if you prefer.

Reply:
But what if someone compiles spin for arm windows and then...erm....deletes rust... 😐

Reply:
Pardon me while I allow that scenario to grip me with terror.